### PR TITLE
feat(problem): generate and compare

### PR DESCRIPTION
**Pull request by tinnamchoi**

### CODE REVIEW

File: crates/cli/src/cli/problem.rs [hunk 1]
```diff
@@ -50,6 +89,25 @@ pub fn cli() -> Command {
...
+                        .action(ArgAction::Set),
+                    Arg::new("lang")
+                        .long("lang")
+                        .help("Language of the first generator file (e.g. cpp, py)")
+                        .action(ArgAction::Set),
+                    Arg::new("problem")
+                        .short('p')
+                        .long("problem")
+                        .help("Problem name (this is not the problem title)")
+                        .action(ArgAction::Set),
```
Comment (rayokamoto) [inline crates/cli/src/cli/problem.rs]: Are there plans to add a second generator file? Otherwise I think "first" can be omitted

File: crates/cli/src/cli/problem.rs [hunk 2]
```diff
@@ -50,6 +89,25 @@ pub fn cli() -> Command {
                         .required(true),
                 ),
         )
+        .subcommand(
+            Command::new("generate")
+                .about("Generate a test case")
```
Comment (rayokamoto) [inline crates/cli/src/cli/problem.rs]: Might want to specify test case input, unless we also allow for the generation of test case outputs (if the user has generator files for both inputs and outputs)

But that output generator is basically the `aucpl problem solve` function so I think we just support test case inputs for now?

File: crates/cli/src/problem/generate.rs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        "generator",
+        generator_file_name,
+        generator_lang,
+        &bin_file,
+    )
+    .context("Failed to get generator command")?;
+
+    let (output, _) = get_output(&bin_file, &script_file, &run_command, None)
+        .context("Failed to get generator output")?;
+    let mut test_file = File::create(problem_path.join("tests/generated.in"))
```
Comment (rayokamoto) [inline crates/cli/src/problem/generate.rs]: I think it'd be nice if the user can also specify the name of the generated file so they don't need to perform the extra step of renaming the generated test file.

I guess they can specify `name` in `tests/{name}.in` - we only allow inputs as, for outputs, the `problem solve` command exists

File: crates/cli/src/problem/generate.rs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+use std::io::Write;
+use std::path::Path;
+
+use anyhow::{Context, Result};
+
+use super::run::{get_cmd, get_output};
+use super::sync_mappings::get_problem;
+use crate::config::Settings;
+use crate::util::get_project_root;
+
```
Comment (rayokamoto) [inline crates/cli/src/problem/generate.rs:11]: Add a short doc comment on what this does, something like

```rs
/// Generate test case inputs based on a generator file
```

File: crates/cli/src/problem/run.rs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+pub fn get_cmd(
+    settings: &Settings,
+    problem: &Path,
+    file_type: &str,
+    file_name: Option<&str>,
+    lang: &String,
+    bin_file: &PathBuf,
+) -> Result<Vec<String>> {
+    let mut file_path = problem.join(format!("{}s/{}.{}", file_type, file_type, lang));
```
Comment (rayokamoto) [inline crates/cli/src/problem/run.rs]: You can actually embed variables in the format strings now, so:

```rs
format!("{file_type}s/{file_type}.{lang}")
```

File: crates/cli/src/problem/run.rs [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+use normpath::PathExt;
+use std::ffi::OsStr;
+use std::fs::{self, File};
+use std::path::{Path, PathBuf};
+use std::time::{Duration, Instant};
+use subprocess::{Exec, Redirection};
+
+use crate::config::Settings;
+
+pub fn get_cmd(
```
Comment (rayokamoto) [inline crates/cli/src/problem/run.rs]: Add function doc comment

I also wonder if we should return something like a `RunCommand` struct with the actual run command as a field, and other metadata such as if there is an associated compiled binary. We can have an implementation of `RunCommand` which can have several methods such as `cleanup()` which will delete compiled binaries if they exist. This doesn't have to be done in this PR, but I think it'll be a good addition.

File: crates/cli/src/problem/run.rs [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+    let run_command = lang_settings.run_command.clone().unwrap_or_default();
+    if run_command.is_empty() {
+        bail!("No run command specified in the settings. It must be specified!");
+    }
+
+    Ok(run_command)
+}
+
+pub fn get_output(
```
Comment (rayokamoto) [inline crates/cli/src/problem/run.rs]: Add function doc comment

It would be good to make the return type a struct as well, as it would make it easier to add new fields later down the line

File: crates/cli/src/problem/compare.rs [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+use anyhow::{Context, Result};
+
+use super::generate;
+use super::run::{get_cmd, get_output};
+use super::sync_mappings::get_problem;
+use crate::util::get_project_root;
+use crate::{config::Settings, util::get_input_files_in_directory};
+
+/// Compare two solutions.
+#[allow(clippy::too_many_arguments)]
```
Comment (rayokamoto) [inline crates/cli/src/problem/compare.rs]: Might be good to have a struct as the argument which can contain all of these parameters as it's getting a bit unwieldy

File: crates/cli/src/problem/compare.rs [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+    )?;
+
+    if *generate {
+        let input_file_path = problem_path.join("tests/generated.in");
+
+        let mut total_tests = 0;
+        let mut total_time_1: f64 = 0f64;
+        let mut total_time_2: f64 = 0f64;
+
+        loop {
```
Comment (rayokamoto) [inline crates/cli/src/problem/compare.rs:47]: As I understand it, this will loop infinitely if there are no mismatches in the two outputs?

File: crates/cli/src/problem/compare.rs [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+    let run_command_2 = get_cmd(
+        settings,
+        &problem_path,
+        "solution",
+        solution_file_name_2,
+        solution_lang_2,
+        &bin_file_2,
+    )?;
+
+    if *generate {
```
Comment (rayokamoto) [inline crates/cli/src/problem/compare.rs:42]: I'm not sure about the utility of this generate option as we have a `problem generate` command.

I guess if the generator command is non-deterministic and the test input is randomised, then I guess having this command makes more sense

File: crates/cli/src/cli/problem.rs [hunk 11]
```diff
@@ -32,6 +32,45 @@ pub fn cli() -> Command {
                         .required(true),
                 ),
         )
+        .subcommand(
```
Comment (rayokamoto) [inline crates/cli/src/cli/problem.rs:39]: Make sure that for the actual arguments (defined via `long()`) that you use kebab-case, rather than snake_case. Using snake_case for the actual argument in `Arg::new` is fine though

**Review by rayokamoto — CHANGES_REQUESTED** [review]

**Review by rayokamoto — CHANGES_REQUESTED** [review]
A few things to address before merging. Some things can be spun off into new tasks though

File: crates/cli/src/problem/compare.rs (hunk 10 above)
Comment (tinnamchoi) [reply crates/cli/src/problem/compare.rs:42]: Yes, the generator is supposed to generate a random test case every time.

File: crates/cli/src/problem/compare.rs (hunk 9 above)
Comment (tinnamchoi) [reply crates/cli/src/problem/compare.rs:47]: Some edge cases may take hundreds or even thousands of random test cases to be discovered, so yes.

File: crates/cli/src/cli/problem.rs (hunk 2 above)
Comment (tinnamchoi) [reply crates/cli/src/cli/problem.rs]: Yeah this is just for inputs

File: crates/cli/src/problem/run.rs [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+        lang: Option<&String>,
+    ) -> Self {
+        Self {
+            name: name.cloned().unwrap_or(format!("{category}")),
+            lang: lang
+                .cloned()
+                .unwrap_or(if category == RunnableCategory::Solution {
+                    settings.problem.default_lang.clone()
+                } else {
+                    settings.problem.default_generator_lang.clone()
```
Comment (rayokamoto) [inline crates/cli/src/problem/run.rs:51]: Make sure to change the `settings.toml.example` file as well if you're adding new command options.
Set a reasonable default for it, whatever language is used the most for generator files.

Make sure to edit the version in that file (to 0.2) and to edit the constant related to that in (I believe) `config.rs`.

File: crates/cli/Cargo.toml [hunk 13]
```diff
@@ -17,4 +17,5 @@ regex.workspace = true
 serde.workspace = true
 serde_json.workspace = true
 subprocess.workspace = true
+uuid = { version = "1.17.0", features = ["v4"] }
```
Comment (rayokamoto) [inline crates/cli/Cargo.toml]: Could you add this to the root Cargo.toml and do `uuid.workspace = true` here?

File: crates/cli/src/problem/compare.rs
Comment (rayokamoto) [inline crates/cli/src/problem/compare.rs:1]: Regarding testing using generators and fuzzing/hacking the solutions to find edge cases - I wonder if this is better suited to a dedicated command? I assumed the `problem compare` command was mostly for comparing two outputs (which is essentially what the `else` branch takes), and I think as it stands, we are combining two commands into one.

I like the idea of a `problem fuzz` to find incorrect answers or edge cases that we haven't considered and implement this very idea into a dedicated command. I think this would allow for more options to be added later down the line as well, and makes the intent of what the command does clearer. 

We can split this out into a new command in a follow-up PR

**Review by rayokamoto — CHANGES_REQUESTED** [review]

**Review by rayokamoto — APPROVED** [review]
